Race condition error while creating basetemp concurrently

I've never seen this error, and can't find a way to reproduce it:

```
10:55:07 INTERNALERROR> Traceback (most recent call last):
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\site-packages\_pytest\main.py", line 203, in wrap_session
10:55:07 INTERNALERROR>     config.hook.pytest_sessionstart(session=session)
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\site-packages\pluggy\hooks.py", line 289, in __call__
10:55:07 INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\site-packages\pluggy\manager.py", line 68, in _hookexec
10:55:07 INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\site-packages\pluggy\manager.py", line 62, in <lambda>
10:55:07 INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\site-packages\pluggy\callers.py", line 208, in _multicall
10:55:07 INTERNALERROR>     return outcome.get_result()
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\site-packages\pluggy\callers.py", line 80, in get_result
10:55:07 INTERNALERROR>     raise ex[1].with_traceback(ex[2])
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\site-packages\pluggy\callers.py", line 187, in _multicall
10:55:07 INTERNALERROR>     res = hook_impl.function(*args)
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\site-packages\xdist\dsession.py", line 81, in pytest_sessionstart
10:55:07 INTERNALERROR>     nodes = self.nodemanager.setup_nodes(putevent=self.queue.put)
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\site-packages\xdist\workermanage.py", line 68, in setup_nodes
10:55:07 INTERNALERROR>     nodes.append(self.setup_node(spec, putevent))
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\site-packages\xdist\workermanage.py", line 77, in setup_node
10:55:07 INTERNALERROR>     node.setup()
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\site-packages\xdist\workermanage.py", line 244, in setup
10:55:07 INTERNALERROR>     basetemp = self.config._tmpdirhandler.getbasetemp()
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\site-packages\_pytest\tmpdir.py", line 117, in getbasetemp
10:55:07 INTERNALERROR>     return py.path.local(self._tmppath_factory.getbasetemp().resolve())
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\site-packages\_pytest\tmpdir.py", line 65, in getbasetemp
10:55:07 INTERNALERROR>     ensure_reset_dir(basetemp)
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\site-packages\_pytest\pathlib.py", line 41, in ensure_reset_dir
10:55:07 INTERNALERROR>     path.mkdir()
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\pathlib.py", line 1248, in mkdir
10:55:07 INTERNALERROR>     self._accessor.mkdir(self, mode)
10:55:07 INTERNALERROR>   File "W:\Miniconda\envs\_dev_eden-win64-py36\lib\pathlib.py", line 387, in wrapped
10:55:07 INTERNALERROR>     return strfunc(str(pathobj), *args)
```

It just happen on the CI server at work. Of course we're using xdist.

I planned to open a PR tomorrow with a reproducer and some working solution for discussion, but here you go:

```python
import os
import stat

def test(tmpdir):
    fn = tmpdir / 'foo.txt'
    fn.write('hello')
    os.chmod(str(fn), stat.S_IREAD)
```

```
λ pytest test-ro.py --basetemp=tmp -q
.
1 passed in 0.02 seconds

λ pytest test-ro.py --basetemp=tmp -q
E
============================== ERRORS ===============================
______________________ ERROR at setup of test _______________________

self = WindowsPath('c:/pytest/.tmp/tmp'), mode = 511, parents = False
exist_ok = False

    def mkdir(self, mode=0o777, parents=False, exist_ok=False):
        """
        Create a new directory at this given path.
        """
        if self._closed:
            self._raise_closed()
        try:
>           self._accessor.mkdir(self, mode)

C:\Users\Bruno\AppData\Local\Programs\Python\Python36\lib\pathlib.py:1246:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

pathobj = WindowsPath('c:/pytest/.tmp/tmp'), args = (511,)

    @functools.wraps(strfunc)
    def wrapped(pathobj, *args):
>       return strfunc(str(pathobj), *args)
E       FileExistsError: [WinError 183] Não é possível criar um arquivo já existente: 'c:\\pytest\\.tmp\\tmp'

C:\Users\Bruno\AppData\Local\Programs\Python\Python36\lib\pathlib.py:387: FileExistsError
====================== short test summary info ======================
ERROR test-ro.py::test - FileExistsError: [WinError 183] Não é poss...
1 error in 0.11 seconds
```

(Note I'm calling pytest twice)

The main issue here is using `--basetemp`, because when that flag is used we will try to use the given directory directly instead of trying to create a numbered dir inside it. That's why I was having this error on just one project, as we don't usually pass `--basetemp` except for that one project.

--- a/changelog/5524.bugfix.rst
+++ b/changelog/5524.bugfix.rst
@@ -1,0 +1,2 @@
+Fix issue where ``tmp_path`` and ``tmpdir`` would not remove directories containing files marked as read-only,
+which could lead to pytest crashing when executed a second time with the ``--basetemp`` option.

--- a/src/_pytest/cacheprovider.py
+++ b/src/_pytest/cacheprovider.py
@@ -14,7 +14,7 @@ import py
 import pytest
 from .pathlib import Path
 from .pathlib import resolve_from_str
-from .pathlib import rmtree
+from .pathlib import rm_rf
 
 README_CONTENT = """\
 # pytest cache directory #
@@ -44,7 +44,7 @@ class Cache:
     def for_config(cls, config):
         cachedir = cls.cache_dir_from_config(config)
         if config.getoption("cacheclear") and cachedir.exists():
-            rmtree(cachedir, force=True)
+            rm_rf(cachedir)
             cachedir.mkdir()
         return cls(cachedir, config)
 

--- a/src/_pytest/pathlib.py
+++ b/src/_pytest/pathlib.py
@@ -7,12 +7,14 @@ import os
 import shutil
 import sys
 import uuid
+import warnings
 from os.path import expanduser
 from os.path import expandvars
 from os.path import isabs
 from os.path import sep
 from posixpath import sep as posix_sep
 
+from _pytest.warning_types import PytestWarning
 
 if sys.version_info[:2] >= (3, 6):
     from pathlib import Path, PurePath
@@ -32,17 +34,42 @@ def ensure_reset_dir(path):
     ensures the given path is an empty directory
     """
     if path.exists():
-        rmtree(path, force=True)
+        rm_rf(path)
     path.mkdir()
 
 
-def rmtree(path, force=False):
-    if force:
-        # NOTE: ignore_errors might leave dead folders around.
-        #       Python needs a rm -rf as a followup.
-        shutil.rmtree(str(path), ignore_errors=True)
-    else:
-        shutil.rmtree(str(path))
+def rm_rf(path):
+    """Remove the path contents recursively, even if some elements
+    are read-only.
+    """
+
+    def chmod_w(p):
+        import stat
+
+        mode = os.stat(str(p)).st_mode
+        os.chmod(str(p), mode | stat.S_IWRITE)
+
+    def force_writable_and_retry(function, p, excinfo):
+        p = Path(p)
+
+        # for files, we need to recursively go upwards
+        # in the directories to ensure they all are also
+        # writable
+        if p.is_file():
+            for parent in p.parents:
+                chmod_w(parent)
+                # stop when we reach the original path passed to rm_rf
+                if parent == path:
+                    break
+
+        chmod_w(p)
+        try:
+            # retry the function that failed
+            function(str(p))
+        except Exception as e:
+            warnings.warn(PytestWarning("(rm_rf) error removing {}: {}".format(p, e)))
+
+    shutil.rmtree(str(path), onerror=force_writable_and_retry)
 
 
 def find_prefixed(root, prefix):
@@ -168,7 +195,7 @@ def maybe_delete_a_numbered_dir(path):
 
         garbage = parent.joinpath("garbage-{}".format(uuid.uuid4()))
         path.rename(garbage)
-        rmtree(garbage, force=True)
+        rm_rf(garbage)
     except (OSError, EnvironmentError):
         #  known races:
         #  * other process did a cleanup at the same time

--- a/testing/test_tmpdir.py
+++ b/testing/test_tmpdir.py
@@ -1,3 +1,5 @@
+import os
+import stat
 import sys
 
 import attr
@@ -311,11 +313,11 @@ class TestNumberedDir:
         )
 
     def test_rmtree(self, tmp_path):
-        from _pytest.pathlib import rmtree
+        from _pytest.pathlib import rm_rf
 
         adir = tmp_path / "adir"
         adir.mkdir()
-        rmtree(adir)
+        rm_rf(adir)
 
         assert not adir.exists()
 
@@ -323,8 +325,39 @@ class TestNumberedDir:
         afile = adir / "afile"
         afile.write_bytes(b"aa")
 
-        rmtree(adir, force=True)
+        rm_rf(adir)
         assert not adir.exists()
+
+    def test_rmtree_with_read_only_file(self, tmp_path):
+        """Ensure rm_rf can remove directories with read-only files in them (#5524)"""
+        from _pytest.pathlib import rm_rf
+
+        fn = tmp_path / "dir/foo.txt"
+        fn.parent.mkdir()
+
+        fn.touch()
+
+        mode = os.stat(str(fn)).st_mode
+        os.chmod(str(fn), mode & ~stat.S_IWRITE)
+
+        rm_rf(fn.parent)
+
+        assert not fn.parent.is_dir()
+
+    def test_rmtree_with_read_only_directory(self, tmp_path):
+        """Ensure rm_rf can remove read-only directories (#5524)"""
+        from _pytest.pathlib import rm_rf
+
+        adir = tmp_path / "dir"
+        adir.mkdir()
+
+        (adir / "foo.txt").touch()
+        mode = os.stat(str(adir)).st_mode
+        os.chmod(str(adir), mode & ~stat.S_IWRITE)
+
+        rm_rf(adir)
+
+        assert not adir.is_dir()
 
     def test_cleanup_ignores_symlink(self, tmp_path):
         the_symlink = tmp_path / (self.PREFIX + "current")
@@ -349,3 +382,24 @@ def attempt_symlink_to(path, to_path):
 
 def test_tmpdir_equals_tmp_path(tmpdir, tmp_path):
     assert Path(tmpdir) == tmp_path
+
+
+def test_basetemp_with_read_only_files(testdir):
+    """Integration test for #5524"""
+    testdir.makepyfile(
+        """
+        import os
+        import stat
+
+        def test(tmp_path):
+            fn = tmp_path / 'foo.txt'
+            fn.write_text('hello')
+            mode = os.stat(str(fn)).st_mode
+            os.chmod(str(fn), mode & ~stat.S_IREAD)
+    """
+    )
+    result = testdir.runpytest("--basetemp=tmp")
+    assert result.ret == 0
+    # running a second time and ensure we don't crash
+    result = testdir.runpytest("--basetemp=tmp")
+    assert result.ret == 0

